modules/keyvault/secret-autorotation/main.bicep

```diff
@@ -0,0 +1,348 @@
+  }
+  dependsOn: [ deployFunctionZip ]
+}]
+
+@description('Create access policy of the keyvault secrets for the function app')
+module addAccessPolicy './modules/access-policy.bicep' = [for item in items(reduce(secrets, {}, (cur, next) => union(cur, { '${next.keyvaultName}': (contains(next, 'keyvaultRg') ? next.keyvaultRg : resourceGroup().name) }))): {
+  name: 'ap-${uniqueString(item.key)}-${suffix}'
+  scope: resourceGroup(item.value)
+  params: {
+    keyvaultName: item.key
+    principalId: functionAppIdentityType == 'SystemAssigned' ? functions.identity.principalId : uai.properties.principalId
+  }
+  dependsOn: [ deployFunctionZip ]
+}]
+
+@batchSize(1)
+@description('Create Keyvault event subscription and let the function app consume it.')
+module eventSubscription './modules/event-subscription.bicep' = [for item in secrets: if (systemTopicName == '') {
+  name: 'kv-event-sub-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(contains(item, 'keyvaultRg') ? item.keyvaultRg : resourceGroup().name)
+  params: {
+    functionAppName: functionAppName
+    functionAppRg: resourceGroup().name
+    functionName: rotationFunctionNames[item.type]
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+  }
+  dependsOn: [
+    grantAccess
+    addAccessPolicy
+  ]
+}]
+
+@batchSize(1)
+@description('Create event subscription to the system topic and let the function app consume it.')
+module topicSubscription './modules/topic-subscription.bicep' = [for item in secrets: if (systemTopicName != '') {
+  name: 'topic-sub-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(systemTopicRg)
+  params: {
+    functionAppName: functionAppName
+    functionAppRg: resourceGroup().name
+    functionName: rotationFunctionNames[item.type]
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+    systemTopicName: systemTopicName
+  }
+  dependsOn: [ deployFunctionZip ]
+}]
+
+@batchSize(1)
+@description('Create keyvault secret if needed. Primary key of the target resource will be used as the initial secret value.')
+module vaultSecrets './modules/create-secret.bicep' = [for item in secrets: if (contains(item, 'isCreate') && item.isCreate) {
+  name: 'kv-secret-${uniqueString('${item.keyvaultName}${item.secretName}')}-${suffix}'
+  scope: resourceGroup(contains(item, 'keyvaultRg') ? item.keyvaultRg : resourceGroup().name)
+  params: {
+    type: item.type
+    resourceName: item.resourceName
+    resourceRg: contains(item, 'resourceRg') ? item.resourceRg : resourceGroup().name
+    keyvaultName: item.keyvaultName
+    secretName: item.secretName
+    validityDays: contains(item, 'validityDays') ? item.validityDays : 90
+    expirationDate: contains(item, 'expirationDate') ? item.expirationDate : dateTimeAdd(now, 'P${contains(item, 'validityDays') ? item.validityDays : 90}D')
```

Comment:
if I rerun the same script twice, will this cause issues since the experationDate will change?

Reply:
That is right, it was provided for the convenience of creating the secret for the first time, but come to think about it again it's better to let the user manage the creation of the key, I'll remove those optional parameters then.